### ISSUE TITLE
Representing opaque structs like `HINSTANCE`

### ISSUE THREAD

**Issue by emirhantasdeviren** [microsoft/windows-rs#1716]
Hello everyone,

I do not know how necessary this is but I will ask about representing opaque types like `HINSTANCE`, `HBRUSH`, `HICON` etc. Any type that is defined with `DECLARE_HANDLE` macro. Here is the definition of the macro in `winnt.h`

```c
#ifdef STRICT
typedef void *HANDLE;
#if 0 && (_MSC_VER > 1000)
#define DECLARE_HANDLE(name) struct name##__; typedef struct name##__ *name
#else
#define DECLARE_HANDLE(name) struct name##__{int unused;}; typedef struct name##__ *name
#endif
#else
typedef PVOID HANDLE;
#define DECLARE_HANDLE(name) typedef HANDLE name
#endif
```
In `STRICT` mode handle types are defined as an opaque struct. [The Rustonomicon](https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs) explains about how to represent opaque structures.

With that I think we can represent handles like this,

```rust
macro_rules! declare_handle {
    ($i:ident) => {
        #[repr(C)]
        pub struct $i {
            _data: [u8; 0],
            _marker: ::core::marker::PhantomData<(*mut u8, ::core::marker::PhantomPinned)>,
        }
    };
}

declare_handle!(HINSTANCE__);
pub type HINSTANCE = *mut HINSTANCE__;
```
I am curious about your thoughts and suggestions.

**Comment by riverar** [comment]
Can you share a bit more as to what you're looking for @emirhantasdeviren? These types are in the crate today.

**Comment by emirhantasdeviren** [comment]
Sorry for not clarifying the purpose of this question.

In the crate `HINSTANCE` is defined as,
```rust
pub type HINSTANCE = isize;
````
This does not provide type safety. In The Rustonomicon way, it provides type safety with not being able to instantiate outside of the module. Also compiler marks the type `!Send`, `!Sync` and `!Unpin`.

**Comment by riverar** [comment]
@emirhantasdeviren Sounds like you're using the raw bindings provided by `windows-sys`. Take a look at the `windows` crate. 

Here's the same type over there https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Foundation/struct.HINSTANCE.html

Also check out ([the FAQ](https://github.com/microsoft/windows-rs/blob/master/docs/FAQ.md#how-do-i-choose-between-the-windows-and-windows-sys-crates) for a bit more info).

**Comment by emirhantasdeviren** [comment]
Yes, I have been looking into `windows-sys`. In my opinion the definition of `HINSTANCE` in `windows` crate can be also be more idiomatic likeso,

```rust
#[repr(C)]
pub struct HINSTANCE__ {
    _data: [u8; 0],
    _marker: ::core::marker::PhantomData<(*mut u8, ::core::marker::PhantomPinned)>,
}

pub struct HINSTANCE(pub ::core::ptr::NonNull<HINSTANCE__>);
```

**Comment by kennykerr** [comment]
Forgive me, but I'm struggling to understand how this is an improvement over the existing definitions.
